File: Napakalaki/src/napakalaki/Napakalaki.java

package napakalaki;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author alex
 */
public class Napakalaki {
    
    // DECLARACIÓN DE VARIABLES
    
    private Monster currentMonster;
    private Player currentPlayer;
    private ArrayList<Player> players;
    private CardDealer dealer;
    
    // CONSTRUCTOR
    
    private Napakalaki() {
        currentMonster = null;
        currentPlayer = null;
        players = new ArrayList();
        dealer = CardDealer.getInstance();
    }
    
    public static Napakalaki getInstance() {
        return NapakalakiHolder.INSTANCE;
    }
    
    private static class NapakalakiHolder {

        private static final Napakalaki INSTANCE = new Napakalaki();
    }
    
    //  MÉTODOS
    
    
    // Inicializa el array de jugadores con los nombres dados.
    
    private void initPlayers(ArrayList<String> names){
        for(String n:names)
            players.add(new Player(n));
    }
    
    
    // Decide qué jugador es el siguiente en jugar.
    
    private Player nextPlayer(){
        int next;
        
        if(currentPlayer == null){
            Random generator = new Random();
            next = generator.nextInt(players.size());
        }
        else{
            if(players.indexOf(currentPlayer) == (players.size()-1))
                next = 0;
            else
                next = players.indexOf(currentPlayer) + 1;
        }
        
        currentPlayer = players.get(next);
        
        return currentPlayer;
    }
    
    
    // Devuelve si el jugador activo puede terminar su turno.
    
    public boolean nextTurnAllowed(){
        return (currentPlayer==null || currentPlayer.validState());
    }
    
    
    // Asigna un enemigo a cada jugador de manera aleatorio y evitando que un
    // jugador sea enemigo de sí mismo.
    
    private void setEnemies(){
        int enemy;
        Random ran = new Random();
        
        for(Player p:players){
            do{
                enemy = ran.nextInt(players.size());
            }while(enemy == players.indexOf(p));
            
            p.setEnemy(players.get(enemy));
        }

    }
    
    
    // Desarrolla la lucha entre el jugador actual y el monstruo actual.
    // Devuelve un objeto de tipo CombatResult.
    
    public CombatResult developCombat(){
        CombatResult combatResult = currentPlayer.combat(currentMonster);
        dealer.giveMonsterBack(currentMonster);
        
        // Código para convertir a Sectario
        
        if(combatResult == CombatResult.LOSEANDCONVERT){
            CultistPlayer p = new CultistPlayer(currentPlayer, dealer.nextCultist());
            int pos = players.indexOf(currentPlayer);
            players.set(pos, p);
            currentPlayer = p;
            
            System.out.println("Ahora eres un jugador sectario.\n");
        }     
        
        return combatResult;
    }
    
    
    // Elimina los tesoros visibles indicados de la lista de tesoros visibles 
    // del jugador. Los tesoros descartados se devuelven al CardDealer.
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        
        for(Treasure t:treasures){
            currentPlayer.discardVisibleTreasure(t);
            dealer.giveTreasureBack(t);
        }
        
    }
    
    
    // Elimina los tesoros ocultos indicados de la lista de tesoros ocultos 
    // del jugador. Los tesoros descartados se devuelven al CardDealer.
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        
        for(Treasure t:treasures){
            currentPlayer.discardHiddenTreasure(t);
            dealer.giveTreasureBack(t);
        }
    
    }
    
    
    // Pasa los tesoros dados a visibles.
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        for(Treasure t:treasures)
            currentPlayer.makeTreasureVisible(t);
    }
    
    
    // Inicializa el juego pidiendo a CardDealer la inicialización de los mazos
    // de cartas, inicializa los jugadores y les asigna sus enemigos. Empieza
    // la ronda asignando el turno.
    
    public void initGame(ArrayList<String> names){
        dealer.initCards();
        initPlayers(names);
        setEnemies();
        nextTurn();
    }
    
    
    // Getter de currentMonster

    public Monster getCurrentMonster() {
        return currentMonster;
    }
    

    // Getter de currentPlayer
    
    public Player getCurrentPlayer() {
        return currentPlayer;
    }
    
    
    // Verifica si el jugador actual está en condiciones de acabar de turno.
    // Si se puede cambiar de turno, se actualiza currentMonster y currentPlayer.
    // En caso de que sea la primera jugada o que el jugador haya muerto se 
    // le reparte tesoros para que pueda jugar.
    
    public boolean nextTurn(){
       boolean stateOK = nextTurnAllowed();
        
        if(stateOK){
            currentMonster = dealer.nextMonster();
            currentPlayer = nextPlayer();
            
            if(currentPlayer.isDead())
                currentPlayer.initTreasures();
        }
        
        return stateOK;
    }
    
    
    // Devuelve true si el parámetro result es WINGAME
    
    public boolean endOfgame(CombatResult result){
        return (result == CombatResult.WINGAME);
    }
    
}   

File: NapakalakiRuby/pkg/NapakalakiRuby-0.0.1/lib/napakalaki.rb
# coding: utf-8

require 'singleton'
require_relative 'card_dealer.rb'
require_relative 'player.rb'
require_relative 'combat_result.rb'
require_relative 'cultist_player.rb'

module Napakalaki
  
  class Napakalaki
    
    include Singleton
    
    # COSULTORES
    
    def getCurrentPlayer
      @currentPlayer
    end
    
    def getCurrentMonster
      @currentMonster
    end
    
    
    # INICIALIZADOR
    
    def initialize
      @currentMonster=nil
      @currentPlayer=nil
      @players=Array.new
      @dealer=CardDealer.instance
    end
    
    # MÉTODOS
    
    private
    
    
    # Inicializa el array de jugadores con el array de nombres dados.
    
    def initPlayers(names)
      names.each do |n|
      @players << Player.new(n)
      end
    end
    
    
    # Decide qué jugador es el siguiente en jugar.
    
    def nextPlayer
      if(@currentPlayer.nil?)
        n=rand(@players.length).to_i
      else
        if(@players.index(@currentPlayer) == (@players.length - 1))
          n=0
        else
          n = @players.index(@currentPlayer) + 1
        end
      end

      @currentPlayer=@players.at(n)

      return @currentPlayer
    end

    
    # Devuelve si el jugador activo puede terminar su turno.
    
    def nextTurnAllowed
      return (@currentPlayer.nil? or @currentPlayer.validState)
    end
    
    
    # Asigna un enemigo a cada jugador de manera aleatorio y evitando que un
    # jugador sea enemigo de sí mismo.

    def setEnemies
      @players.each do |p|
        begin
          pos = Random.rand(@players.size)
        end while(pos == @players.index(p))
        
        enemy = @players.at(pos);
        p.setEnemy(enemy)
      end
    end
    
    
    # Desarrolla la lucha entre el jugador actual y el monstruo actual.
    # Devuelve un objeto de tipo CombatResult.
    
    public
    def developCombat
      combat_result=@currentPlayer.combat(@currentMonster)
      @dealer.giveMonsterBack(@currentMonster)
      
      # Código para convertir a sectario
      
      if(combat_result==CombatResult::LOSEANDCONVERT)
        p=CultistPlayer.new(@currentPlayer, @dealer.nextCultist)
        pos=@players.index(@currentPlayer)
        @players[pos]=p
        @currentPlayer=p
        
        puts "Ahora eres un jugador sectario.\n"
      end
      
      return combat_result
    end
    
    
    # Elimina los tesoros ocultos indicados de la lista de tesoros ocultos 
    # del jugador. Los tesoros descartados se devuelven al CardDealer.
    
    def discardVisibleTreasures(t)
      
      t.each do |x|
        @currentPlayer.discardVisibleTreasure(x)
        @dealer.giveTreasureBack(x)
      end
      
    end
    
    
    # Elimina los tesoros ocultos indicados de la lista de tesoros ocultos 
    # del jugador. Los tesoros descartados se devuelven al CardDealer.
    
    
    def discardHiddenTreasures(t)
      
      t.each do |x|
        @currentPlayer.discardHiddenTreasure(x)
        @dealer.giveTreasureBack(x)
        
      end
      
    end
    
    
    # Pasa todos los tesoros dados a visibles(equipados).
    
    def makeTreasuresVisible(t)
      
      t.each do |x|
        @currentPlayer.makeTreasureVisible(x)
      end
      
    end
    
    
    # Inicializa el juego pidiendo a CardDealer la inicialización de los mazos
    # de cartas, inicializa los jugadores y les asigna sus enemigos. Empieza
    # la ronda asignando el turno.
      
    def initGame(players)
      @dealer.initCards
      initPlayers(players)
      setEnemies
      nextTurn
    end
    
    
    # Verifica si el jugador actual está en condiciones de acabar de turno.
    # Si se puede cambiar de turno, se actualiza currentMonster y currentPlayer.
    # En caso de que sea la primera jugada o que el jugador haya muerto se 
    # le reparte tesoros para que pueda jugar.
    
    def nextTurn
      state_ok=nextTurnAllowed

      if(state_ok)
        @currentMonster=@dealer.nextMonster
        @currentPlayer=nextPlayer
        
        if(@currentPlayer.isDead)
          @currentPlayer.initTreasures
        end
      end
      
      return state_ok
    end
    
    
    # Devuelve true si el parámetro result es WINGAME
    
    def endOfGame(result)
      return CombatResult::WINGAME==result
    end
 
  end
end